query/language/fetch.feature

```diff
@@ -11,855 +11,2942 @@ Feature: TypeQL Fetch Query
+          }
+        ]
+      }
+      """
+    Then answer contains document:
+      """
+      {
+        "list pipeline": [ ]
+      }
+      """
+
+    When transaction closes
+    When connection open write transaction for database: typedb
+    When typeql write query
+      """
+      match
+        $p isa person, has person-name "Alice", has person-name "Allie";
+      insert
+        $p has person-name "Alicia";
+      """
+    When transaction commits
+    When connection open read transaction for database: typedb
+
+
+    When get answers of typeql read query
+      """
+      match
+        $p isa person, has person-name $n;
+      fetch {
+        "list pipeline": [
+            match
+              $p has person-name $pn;
+              $pn != $n;
+            fetch {
+              "name": $n,
+              "also known as": $pn
+            };
+        ],
+      };
+      """
+    Then answer size is: 4
+    Then answer contains document:
+      """
+      {
+        "list pipeline": [
+          {
+            "name": "Alice",
+            "also known as": "Allie"
+          },
+          {
+            "name": "Alice",
+            "also known as": "Alicia"
+          }
+        ]
+      }
+      """
+    Then answer contains document:
+      """
+      {
+        "list pipeline": [
+          {
+            "name": "Allie",
+            "also known as": "Alice"
+          },
+          {
+            "name": "Allie",
+            "also known as": "Alicia"
+          }
+        ]
+      }
+      """
+    Then answer contains document:
+      """
+      {
+        "list pipeline": [
+          {
+            "name": "Alicia",
+            "also known as": "Alice"
+          },
+          {
+            "name": "Alicia",
+            "also known as": "Allie"
+          }
+        ]
+      }
+      """
+    Then answer contains document:
+      """
+      {
+        "list pipeline": [ ]
+      }
+      """
+
+
+  Scenario: same fetch parameter names in one fetch object are not permitted
+    Then typeql read query; fails with a message containing: "multiple mappings for one key"
+      """
+      match
+      $p isa! $t, has person-name $n;
+      fetch {
+        "first": $t,
+        "first": $n,
+      };
+      """
+    Then typeql read query; fails with a message containing: "multiple mappings for one key"
+      """
+      match
+      $p isa! $t, has person-name $n;
+      fetch {
+        "first": $t,
+        "first": [ $p.ref ],
+      };
+      """
+    Then typeql read query; fails with a message containing: "multiple mappings for one key"
+      """
+      match
+      $p isa! $t, has person-name $n;
+      fetch {
+        "first": $n,
+        "first": [ $p.person-name ],
+      };
+      """
+    Then typeql read query; fails with a message containing: "multiple mappings for one key"
+      """
+      match
+      $p isa! $t, has person-name $n;
+      fetch {
+        "first": $n,
+        "first": $n,
+      };
```

Comment:
For future reference, this is great - but I think you don't really have to write this level of granularity for tests. I would generally consider: is the sensible data structure/check going to be consistently implemented the same way in all these scenarios (eg. the keys are stored in a map and the same logic is used in all the cases)? If yes, then the tests don't need to be duplicated. This isn't _super_ comprehensive and relies on everyone coding in a good and similar way, but it is a level aim to achieve and reduces the number of tests we have to write & maintain (at risk of missing something sometimes).

Reply:
I'd aim to be granular in our tests. We are creating a database, and, where possible, it would be great to have the behavior pinned in tests, so we don't miss anything.

I agree that I may be sometimes too granular in my checks and worries, so feel free to continue correcting me, please. Although in this specific case... I spent 15 minutes debugging why my tests were not working, and then I noticed that I had two headers duplicated and squeezed into a single json document entry. I imagined that it would be frustrating for our users and, considering that we are a statically strong database, I decided to add this check, and I expect it to be constant, and this test highlights that we **do** care about this behavior. 

So I did it partially for myself. And I think that I'm correct here

Reply:
Fair enough I don't mind it - just keep in mind the overall tradeoff of speed vs accuracy. We definitely want to cover these edge cases at least once.